### ISSUE TITLE
Ideas

### ISSUE THREAD

**Issue by ryasmi** [ryansmith94/rulr#47]
- API Interface for documentation and clarity in breaking changes?
- Better documentation.
- Reduce API surface
  - Remove functions that create warnings and just return the warning instances.
  - Remove  under-used functions `checkRegex`, `checkType`, `checkThrow`, and `checkBool`.
- Improve API
  - Add `matches` and `between` to go with `nullable`, `required`, and `optional`.
- Improve naming?
  - Replace `first` with `sequence`.
  - Replace `composeRules` with `all`?
  - Replace `restrictToSchema` with `schema`?
  - Replace `restrictToCollection` with `collection`?
  - Replace `hasSchema` with ...?
  - Replace `restrictToKeys` with ...?

Ideally I'd like the rules to read correctly in the code.

```ts
validateData(hasObjectWhere({
  x: either(hasFloat, hasNull, hasUndefined), // x either has float, has null, (or) has undefined.
  y: hasString,
  z: only([
    hasObjectWhere({
      a: hasArrayWhere(item => hasFloat),
      b: isMatching('hello'),
      c: firstly(hasInteger, isBetween(0, 100)), // c firstly has integer (that) is between 0 (and) 100.
    }),
    hasFirstAMatchingC,
  ]), // z only has object where ... (and) has first a matching c.
}));
```

Potential New API
```ts
Rule: <D>(data: D, path: string) => ValidationError[]
Schema: { [key: string]: Rule }
validateData: <D>(rule: Rule<D>) => (data: D, path: string[]) => D // throws ValidationErrors
hasObjectWhere: (schema: Schema) => Rule // compound/reference rule
hasArrayWhere: ((item) => Rule) => Rule // compound/reference rule
hasString: Rule<string> // scalar rule
hasInteger: Rule<number> // scalar rule
hasFloat: Rule<number> // scalar rule
hasBoolean: Rule<boolean> // scalar rule
hasNull: Rule<null> // scalar rule
hasUndefined: Rule<undefined> // scalar rule
isBetween: (min = -Infinity, max = Infinity) => Rule<number> // conditional rule
isMatching: <V>(value: V) => Rule<V> // condition rule
firstly: <V>(...rules: Rule<V>[]) => Rule<V> // composer rule - sequential all
only: <V>(...rules: Rule<V>[]) => Rule<V> // composer rule - parallel all
either: <V>(...rules: Rule<V>[]) => Rule<V> // composer rule - parallel some
```

Should decide whether `required` or `optional` is superfluous. Required is probably superfluous because `undefined` is not a number, but `undefined` is an optional number.

Might be cool if it created a type based on the validation.

**Comment by ryansmith94** [comment]
This actually allows type checking. Pretty awesome.

```ts
type ValidationError = string;
type Rule<D> = (data: D) => ValidationError[];
type UnruleProp<T> = T extends Rule<infer V> ? V : T;
type Schema<T> = {
  readonly [P in keyof T]: Rule<T[P]>;
}
type Unschema<T> = {
  readonly [P in keyof T]: UnruleProp<T[P]>;
}
type ValidateData = <D>(rule: Rule<D>) => (data: D) => D; // throws ValidationErrors
type HasObjectWhere = <T, S extends Schema<T>>(schema: S) => Rule<Unschema<S>>; // compound/reference rule
type HasString = Rule<string> // scalar rule
type HasInteger = Rule<number> // scalar rule
type HasFloat = Rule<number> // scalar rule
type HasBoolean = Rule<boolean> // scalar rule
type HasNull = Rule<null> // scalar rule
type HasUndefined = Rule<undefined> // scalar rule
type IsBetween = (min: number, max: number) => Rule<number> // conditional rule
type IsMatching = <V>(value: V) => Rule<V> // condition rule
type Firstly = <V>(...rules: Rule<V>[]) => Rule<V> // composer rule - sequential all
type Only = <V>(...rules: Rule<V>[]) => Rule<V> // composer rule - parallel all
type Either = <V>(...rules: Rule<V>[]) => Rule<V> // composer rule - parallel some

const firstly: Firstly = (...rules) => (data) => {
  return rules.reduce((errors, rule) => {
    if (errors.length === 0) {
      return rule(data);
    }
    return errors;
  }, []);
};

const either: Either = (...rules) => (data) => {
  const errorsOfRules = rules.map((rule) => {
    return rule(data);
  })
  const hasRuleWithoutErrors = errorsOfRules.filter((errors) => {
    return errors.length === 0;
  }).length !== 0;

  if (hasRuleWithoutErrors) {
    return [];
  }
  return [].concat(...errorsOfRules);
};

const hasObjectWhere: HasObjectWhere = (schema) => (data) => {
  return Object.keys(schema).reduce((errors, key) => {
    const rule = schema[key];
    return [...errors, rule(data[key])];
  }, []);
};

const isBetween: IsBetween = (min, max) => (data) => {
  if (min < data && data < max) {
    return [];
  }
  return ['not between min and max'];
};

const isMatching: IsMatching = (value) => (data) => {
  if (data === value) {
    return [];
  }
  return ['not matching value'];
}

const hasString: HasString = (data) => {
  if (typeof data === 'string') {
    return [];
  }
  return ['not a string'];
};

const hasInteger: HasInteger = (data) => {
  if (typeof data === 'number' && Math.floor(data) - data === data) {
    return [];
  }
  return ['not an integer'];
};

const validate: ValidateData = (rule) => (data) => {
  const errors = rule(data);
  if (errors.length === 0) {
      return data;
  }
  throw new Error(errors.join('\n'));
};

const myValidData = validate(hasObjectWhere({
    x: hasInteger,
    y: hasObjectWhere({
        z: either<string | number>(hasString, hasInteger),
    }),
}))({ x: 10, y: { z: 10 } });
```

**Comment by ryansmith94** [comment]
Had been working on this in the v5 branch but now deprecating this repository in favour of https://github.com/pelotom/runtypes

**Comment by ryansmith94** [comment]
Decided to continue this package due to runtype not providing all errors.

**Comment by ryansmith94** [comment]
Closed by #51.
